src/coprocessor/dag/rpn_expr/types.rs

```diff
@@ -0,0 +1,695 @@
+        match self {
+            RpnExpressionNode::Constant { ref value, .. } => Some(value),
+            _ => None,
+        }
+    }
+
+    /// Gets the column offset for `TableColumnRef` variant.
+    #[inline]
+    pub fn table_column_ref_offset(&self) -> Option<usize> {
+        match self {
+            RpnExpressionNode::TableColumnRef { ref offset, .. } => Some(*offset),
+            _ => None,
+        }
+    }
+}
+
+/// An RPN expression node list which represents an expression in Reverse Polish notation.
+#[derive(Debug)]
+pub struct RpnExpressionNodeVec(Vec<RpnExpressionNode>);
+
+impl std::ops::Deref for RpnExpressionNodeVec {
+    type Target = Vec<RpnExpressionNode>;
+
+    fn deref(&self) -> &Vec<RpnExpressionNode> {
+        &self.0
+    }
+}
+
+impl std::ops::DerefMut for RpnExpressionNodeVec {
+    fn deref_mut(&mut self) -> &mut Vec<RpnExpressionNode> {
+        &mut self.0
+    }
+}
+
+impl RpnExpressionNodeVec {
+    /// Builds the RPN expression node list from an expression definition tree.
+    // TODO: Deprecate it in Coprocessor V2 DAG interface, since we should accept RPN list directly.
+    pub fn build_from_expr_tree(def: Expr, time_zone: Tz) -> Result<Self> {
+        let mut expr_nodes = Vec::new();
+        Self::append_rpn_nodes_recursively(
+            def,
+            time_zone,
+            &mut expr_nodes,
+            super::map_pb_sig_to_rpn_func,
+        )?;
+        Ok(RpnExpressionNodeVec(expr_nodes))
+    }
+
+    /// Transforms eval tree nodes into RPN nodes.
+    ///
+    /// Suppose that we have a function call:
+    ///
+    /// ```ignore
+    /// A(B, C(E, F, G), D)
+    /// ```
+    ///
+    /// The eval tree looks like:
+    ///
+    /// ```ignore
+    ///           +---+
+    ///           | A |
+    ///           +---+
+    ///             |
+    ///   +-------------------+
+    ///   |         |         |
+    /// +---+     +---+     +---+
+    /// | B |     | C |     | D |
+    /// +---+     +---+     +---+
+    ///             |
+    ///      +-------------+
+    ///      |      |      |
+    ///    +---+  +---+  +---+
+    ///    | E |  | F |  | G |
+    ///    +---+  +---+  +---+
+    /// ```
+    ///
+    /// We need to transform the tree into RPN nodes:
+    ///
+    /// ```ignore
+    /// B E F G C D A
+    /// ```
+    ///
+    /// The transform process is very much like a post-order traversal. This function does it
+    /// recursively.
+    fn append_rpn_nodes_recursively<F>(
+        mut def: Expr,
+        time_zone: Tz,
+        rpn_nodes: &mut Vec<RpnExpressionNode>,
+        fn_mapper: F,
+    ) -> Result<()>
+    where
+        F: Fn(tipb::expression::ScalarFuncSig) -> Result<Box<dyn RpnFunction>> + Copy,
+    {
+        // TODO: We should check whether node types match the function signature. Otherwise there
+        // will be panics when the expression is evaluated.
+
+        use crate::coprocessor::codec::mysql::{Decimal, Duration, Json, Time, MAX_FSP};
```

Comment:
Could we move `use ..`  to the front of this file then we could just focus on the main logic

Reply:
You find it! XD

It is specially designed to be placed here, locally, i.e. inside this function, because the usage of these modules (decimals, durations, etc) only (and should only) exists inside this function. In other words, they are not globally, or widely used outside, only locally used. So they are only local imports. This is to minimize the possibility and work of re-arranging import items when we move this functionality in future.

In contrast, for globally / widely used modules, like Error and Result, they are globally imported.

Reply:
This is similar to how one uses Diesel then?

Reply:
What is it 😆?